sdk/detect/env.go

```diff
@@ -0,0 +1,89 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package env provides a detector that loads resource information from
+// the OTEL_RESOURCE environment variable. A list of labels of the form
+// `<key1>=<value1>,<key2>=<value2>,...` is accepted. Domain names and
+// paths are accepted as label keys.
+package detect
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"regexp"
+	"strings"
+
+	"go.opentelemetry.io/otel/api/kv"
+	"go.opentelemetry.io/otel/sdk/resource"
+)
+
+// Environment variable used by "env" to decode a resource.
+const envVar = "OTEL_RESOURCE"
+
+// FromEnv is a detector that implments the ResourceDetector and collects resources
+// from environment
+type FromEnv struct{}
+
+// Detect is a function that collects resources from environment
+func (d *FromEnv) Detect(context.Context) (*resource.Resource, error) {
+	labels := strings.TrimSpace(os.Getenv(envVar))
+
+	if labels == "" {
+		return resource.Empty(), nil
+	}
+	return constructOTResources(labels)
+}
+
+// labelRegex matches any key=value pair including a trailing comma or the end of the
+// string. Captures the trimmed key & value parts, and ignores any superfluous spaces.
+var labelRegex = regexp.MustCompile(`\s*([[:ascii:]]{1,256}?)\s*=\s*([[:ascii:]]{0,256}?)\s*(?:,|$)`)
+
+func constructOTResources(s string) (*resource.Resource, error) {
+	matches := labelRegex.FindAllStringSubmatchIndex(s, -1)
```

Comment:
Is a regex necessary here?  It feels overkill.  I think the key=value pairs can be handled by first splitting on `,` and then on `=`, as is done in the [Jaeger exporter's tag extractor](https://github.com/open-telemetry/opentelemetry-go/blob/master/exporters/trace/jaeger/env.go#L115).

Reply:
Agree. String operations are more effective than regex.